cpp/src/join/mixed_join_semi.cu

```diff
@@ -328,243 +249,36 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
-      case join_kind::LEFT_SEMI_JOIN: return {0, std::move(matches_per_row)};
-      default: CUDF_FAIL("Invalid join kind."); break;
-    }
-  } else if (left_num_rows == 0) {
-    switch (join_type) {
-      // Left, left anti, left semi, and inner joins all return empty sets.
-      case join_kind::LEFT_ANTI_JOIN:
-      case join_kind::LEFT_SEMI_JOIN: {
-        thrust::fill(matches_per_row->begin(), matches_per_row->end(), 0);
-        return {0, std::move(matches_per_row)};
-      }
-      default: CUDF_FAIL("Invalid join kind."); break;
-    }
-  }
-
-  // If evaluating the expression may produce null outputs we create a nullable
-  // output column and follow the null-supporting expression evaluation code
-  // path.
-  auto const has_nulls = cudf::nullate::DYNAMIC{
-    cudf::has_nulls(left_equality) || cudf::has_nulls(right_equality) ||
-    binary_predicate.may_evaluate_null(left_conditional, right_conditional, stream)};
-
-  auto const parser = ast::detail::expression_parser{
-    binary_predicate, left_conditional, right_conditional, has_nulls, stream, mr};
-  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
-               "The expression must produce a boolean output.");
+  auto gather_map = std::make_unique<rmm::device_uvector<size_type>>(probe.num_rows(), stream, mr);
 
-  // TODO: The non-conditional join impls start with a dictionary matching,
-  // figure out what that is and what it's needed for (and if conditional joins
-  // need to do the same).
-  auto& probe                 = swap_tables ? right_equality : left_equality;
-  auto& build                 = swap_tables ? left_equality : right_equality;
-  auto probe_view             = table_device_view::create(probe, stream);
-  auto build_view             = table_device_view::create(build, stream);
-  auto left_conditional_view  = table_device_view::create(left_conditional, stream);
-  auto right_conditional_view = table_device_view::create(right_conditional, stream);
-
-  auto const preprocessed_build =
-    experimental::row::equality::preprocessed_table::create(build, stream);
-  auto const preprocessed_probe =
-    experimental::row::equality::preprocessed_table::create(probe, stream);
-  auto const row_comparator =
-    cudf::experimental::row::equality::two_table_comparator{preprocessed_probe, preprocessed_build};
-  auto const equality_probe = row_comparator.equal_to<false>(has_nulls, compare_nulls);
+  // gather_map_end will be the end of valid data in gather_map
+  auto gather_map_end =
+    thrust::copy_if(rmm::exec_policy(stream),
+                    thrust::counting_iterator<size_type>(0),
+                    thrust::counting_iterator<size_type>(probe.num_rows()),
+                    gather_map->begin(),
+                    [join_type, d_flagged = flagged->begin()] __device__(size_type const idx) {
+                      return *(d_flagged + idx) == (join_type == detail::join_kind::LEFT_SEMI_JOIN);
+                    });
```

Comment:
*suggestion* which may not be better, rather than closing over the flagged iterator in the lambda, how about using the `stencil` form of `copy_if`

```
auto gather_map_end =
  thrust::copy_if(rmm::exec_policy(stream),
                  thrust::counting_iterator<size_type>(0),
                  thrust::counting_iterator<size_type>(probe.num_rows()),
                  flagged.begin(),
                  gather_map->begin(),
                  [join_type] __device__(bool flag) {
                    return flag == (join_type == detail::join_kind::LEFT_SEMI_JOIN);
                  });
```
WDYT?

Reply:
fixed, I think this is better, thanks!